Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp

```diff
@@ -0,0 +1,389 @@
+ /* Copyright 2024 The WarpX Community
+ *
+ * This file is part of WarpX.
+ *
+ * Authors: S. Eric Clark (Helion Energy, Inc.)
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+
+#include "ProjectionDivCleaner.H"
+
+#include <AMReX_MLPoisson.H>
+#include <AMReX_MultiFabUtil.H>
+#include <AMReX_LO_BCTYPES.H>
+
+#include <WarpX.H>
+#if defined WARPX_DIM_RZ
+#include <FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CylindricalYeeAlgorithm.H>
+#else
+#include <FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CartesianYeeAlgorithm.H>
+#endif
+#include <FieldSolver/Fields.H>
+#include <Initialization/ExternalField.H>
+#include <ablastr/utils/Communication.H>
+#include <Utils/WarpXProfilerWrapper.H>
+
+using namespace amrex;
+
+namespace warpx::initialization {
+
+ProjectionDivCleaner::ProjectionDivCleaner(warpx::fields::FieldType a_field_type)
+{
+    m_field_type = a_field_type;
+
+    // Initialize tolerance based on field precision
+    if constexpr (std::is_same<Real, float>::value) {
+        m_rtol = 5e-5;
+        m_atol = 0.0;
+    }
+    else {
+        m_rtol = 1e-12;
+        m_atol = 0.0;
+    }
+
+    auto& warpx = WarpX::GetInstance();
+    m_levels = warpx.finestLevel() + 1;
+
+    m_solution.resize(m_levels);
+    m_source.resize(m_levels);
+
+    const int ncomps = WarpX::ncomps;
+    auto const& ng = warpx.getFieldPointer(m_field_type, 0, 0)->nGrowVect();
+
+    std::array<amrex::Real,3> const& cell_size = WarpX::CellSize(0);
+
+    for (int lev = 0; lev < m_levels; ++lev)
+    {
+        // Default BoxArray and DistributionMap for initializing the output MultiFab, m_mf_output.
+        const amrex::BoxArray& ba = warpx.boxArray(lev);
+        const amrex::DistributionMapping& dmap = warpx.DistributionMap(lev);
+
+        m_solution[lev].reset();
+        m_source[lev].reset();
+
+        const auto tag1 = amrex::MFInfo().SetTag("div_cleaner_solution");
+        m_solution[lev] = std::make_unique<MultiFab>(amrex::convert(ba, IntVect::TheCellVector()),
+            dmap, ncomps, ng, tag1);
+        const auto tag2 = amrex::MFInfo().SetTag("div_cleaner_source");
+        m_source[lev] = std::make_unique<MultiFab>(amrex::convert(ba, IntVect::TheCellVector()),
+            dmap, ncomps, ng, tag2);
+
+        m_solution[lev]->setVal(0.0, ng);
+        m_source[lev]->setVal(0.0, ng);
+    }
+
+    m_h_stencil_coefs_x.resize(1);
```

Comment:
Could this routine be used here? `CylindricalYeeAlgorithm::InitializeStencilCoefficients`

Reply:
I think it could be used here. I will attempt to do a refactor.